diff --git a/libusb/os/darwin_usb.c b/libusb/os/darwin_usb.c
index 204e1ef..c6f2f67 100644
--- a/libusb/os/darwin_usb.c
+++ b/libusb/os/darwin_usb.c
@@ -623,7 +623,13 @@ static int darwin_cache_device_descriptor (struct libusb_context *ctx, struct da
   /* According to Apple's documentation the device must be open for DeviceRequest but we may not be able to open some
    * devices and Apple's USB Prober doesn't bother to open the device before issuing a descriptor request.  Still,
    * to follow the spec as closely as possible, try opening the device */
-  is_open = ((*device)->USBDeviceOpenSeize(device) == kIOReturnSuccess);
+  if (kIOUSBVendorIDAppleComputer != idVendor) {
+    is_open = ((*device)->USBDeviceOpenSeize(device) == kIOReturnSuccess);
+  } else {
+    // don't open Apple devices, brickd is not interested in them, and opening an Apple USB-Ethernet-Adapter
+    // breaks the Ethernet connection of this adapter
+    is_open = 0;
+  }
 
   do {
     /**** retrieve device descriptor ****/
